crates/install-wheel-rs/src/linker.rs

```diff
@@ -280,55 +281,118 @@ fn clone_wheel_files(
+                        if reflink::reflink(&from, &tempfile).is_ok() {
+                            fs::rename(&tempfile, to)?;
+                        } else {
+                            debug!("Failed to clone {} to temporary location {} - attempting to copy files as a fallback", from.display(), tempfile.display());
+                            *attempt = Attempt::UseCopyFallback;
+                            fs::copy(&from, &to)?;
+                        }
+                    }
+                } else {
+                    debug!(
+                        "Failed to clone {} to {} - attempting to copy files as a fallback",
+                        from.display(),
+                        to.display()
+                    );
+                    // switch to copy fallback
+                    *attempt = Attempt::UseCopyFallback;
+                    clone_recursive(site_packages, wheel, entry, attempt)?;
+                }
+            }
+        }
+        Attempt::Subsequent => {
+            if let Err(err) = reflink::reflink(&from, &to) {
+                if matches!(err.kind(), std::io::ErrorKind::AlreadyExists) {
+                    // If cloning/copying fails and the directory exists already, it must be merged recursively.
+                    if entry.file_type()?.is_dir() {
+                        for entry in fs::read_dir(from)? {
+                            clone_recursive(site_packages, wheel, &entry?, attempt)?;
+                        }
+                    } else {
+                        // If file already exists, overwrite it.
+                        let tempdir = tempdir_in(site_packages)?;
+                        let tempfile = tempdir.path().join(from.file_name().unwrap());
+                        reflink::reflink(&from, &tempfile)?;
+                        fs::rename(&tempfile, to)?;
+                    }
+                } else {
+                    return Err(Error::Reflink { from, to, err });
+                }
+            }
+        }
+        Attempt::UseCopyFallback => {
+            if entry.file_type()?.is_dir() {
+                fs::create_dir_all(&to)?;
+                for entry in fs::read_dir(from)? {
+                    clone_recursive(site_packages, wheel, &entry?, attempt)?;
+                }
+            } else {
+                fs::copy(&from, &to)?;
             }
-        } else {
-            // If file already exists, overwrite it.
-            let tempdir = tempdir_in(site_packages)?;
-            let tempfile = tempdir.path().join(from.file_name().unwrap());
-            reflink_copy::reflink(from, &tempfile)?;
-            fs::rename(&tempfile, to)?;
         }
-    } else {
-        // Other errors should be tracked
-        reflink.map_err(|err| Error::Reflink {
-            from: from.clone(),
-            to: to.clone(),
-            err,
-        })?;
     }
 
+    if *attempt == Attempt::Initial {
+        *attempt = Attempt::Subsequent;
+    }
```

Comment:
Is this different than including it as an `else` in the `Attempt::Initial` arm?

E.g.:

```diff
diff --git a/crates/install-wheel-rs/src/linker.rs b/crates/install-wheel-rs/src/linker.rs
index 6275994f..69eff653 100644
--- a/crates/install-wheel-rs/src/linker.rs
+++ b/crates/install-wheel-rs/src/linker.rs
@@ -356,6 +356,9 @@ fn clone_recursive(
                     *attempt = Attempt::UseCopyFallback;
                     clone_recursive(site_packages, wheel, entry, attempt)?;
                 }
+            } else {
+                // If the first reflink succeeded, we'll use reflink for the rest of the install.
+                *attempt = Attempt::Subsequent;
             }
         }
         Attempt::Subsequent => {
@@ -390,9 +393,6 @@ fn clone_recursive(
         }
     }

-    if *attempt == Attempt::Initial {
-        *attempt = Attempt::Subsequent;
-    }
     Ok(())
 }
```

Reply:
It would be, because we also want to update the attempt value in the case of an `AlreadyExists` error that we handle. For example, if we have to overwrite a file, but don't need to fall-back to copying, this `else` branch would not be hit.